xls/modules/zstd/buffer.x

```diff
@@ -0,0 +1,451 @@
+    assert_eq(buffer_has_at_least(buffer, u32:16), true);
+    assert_eq(buffer_has_at_least(buffer, u32:32), false);
+    assert_eq(buffer_has_at_least(buffer, u32:33), false);
+
+    let buffer = Buffer { contents: u32:0, length: u32:32 };
+    assert_eq(buffer_has_at_least(buffer, u32:0), true);
+    assert_eq(buffer_has_at_least(buffer, u32:16), true);
+    assert_eq(buffer_has_at_least(buffer, u32:32), true);
+    assert_eq(buffer_has_at_least(buffer, u32:33), false);
+}
+
+// Appends `data` to a `buffer`. Returns the updated buffer
+// It will fail if the buffer cannot fit the data. For calls that need
+// better error handling, check `buffer_append_checked`
+pub fn buffer_append<BSIZE: u32, DSIZE: u32> (buffer: Buffer<BSIZE>, data: bits[DSIZE]) -> Buffer<BSIZE> {
+    if buffer_can_fit(buffer, data) == false {
+        trace_fmt!("Not enough space in the buffer! {} + {} <= {}", buffer.length, DSIZE, BSIZE);
+        fail!("not_enough_space", ());
+    } else {
+        ()
+    };
+
+    Buffer {
+        contents: (data as bits[BSIZE] << buffer.length) | buffer.contents,
+        length: DSIZE + buffer.length
+    }
+}
+
+#[test]
+fn test_buffer_append() {
+    let buffer = Buffer { contents: u32:0, length: u32:0 };
+    let buffer = buffer_append(buffer, u16:0xBEEF);
+    assert_eq(buffer, Buffer { contents: u32:0xBEEF, length: u32:16 });
+    let buffer = buffer_append(buffer, u16:0xDEAD);
+    assert_eq(buffer, Buffer { contents: u32:0xDEADBEEF, length: u32:32 });
+}
+
+// Appends `data` to a `buffer` if possible. Returns the status of the operation
+// and the updated buffer in a BufferResult structure
+pub fn buffer_append_checked<BSIZE: u32, DSIZE: u32> (buffer: Buffer<BSIZE>, data: bits[DSIZE]) -> BufferResult<BSIZE> {
+    if buffer_can_fit(buffer, data) == false {
+        BufferResult { status: BufferStatus::NO_ENOUGH_SPACE, buffer }
+    } else {
+        BufferResult {
+            status: BufferStatus::OK,
+            buffer: Buffer {
+                contents: (data as bits[BSIZE] << buffer.length) | buffer.contents,
+                length: DSIZE + buffer.length
+            }
+        }
+    }
+}
+
+#[test]
+fn test_buffer_append_checked() {
+    let buffer = Buffer { contents: u32:0, length: u32:0 };
+
+    let result1 = buffer_append_checked(buffer, u16:0xBEEF);
+    assert_eq(result1, BufferResult {
+        status: BufferStatus::OK,
+        buffer: Buffer { contents: u32:0xBEEF, length: u32:16 }
+    });
+
+    let result2 = buffer_append_checked(result1.buffer, u16:0xDEAD);
+    assert_eq(result2, BufferResult {
+        status: BufferStatus::OK,
+        buffer: Buffer { contents: u32:0xDEADBEEF, length: u32:32 }
+    });
+
+    let result3 = buffer_append_checked(result2.buffer, u16:0xCAFE);
+    assert_eq(result3, BufferResult {
+        status: BufferStatus::NO_ENOUGH_SPACE,
+        buffer: result2.buffer
+    });
+}
+
+// Pops `length` amount of data from a `buffer`. Returns the modified buffer
```

Comment:
should we specify that the data is `pop'ed` in chronological order (fifo)

Reply:
I updated the Buffer description at the beginning of the file and also in this doc-string